### ISSUE TITLE
take effect for SIGNUP event

### ISSUE THREAD

**Issue by Puspendert** [wtjs/what-the-splash#1]
Hello,
It's not an issue, just a query I want to ask.
In Lesson-5, you very well explained about the `take` effect using the LOGIN LOGOUT events.  Now I am trying to applying the same login to my SIGNUP action.
User press signup button, SIGNUP event gets fired ->  it could pass or fail -> if pass then fire SIGNUP_SUCCESS and then WELCOME_NOTE -> if fail then fire SIGNUP_FAIL.

```
function* watchSignupSaga(){
  yield take(NEW_USER.SIGNUP);
  call(handleSignup)    

  yield take(NEW_USER.SIGNUP_SUCCESS);
  call(handleUserSignupSucces)

  yield take(NEW_USER.WELCOME_NOTE)
  call(handleWelcomeNote)

  yield take(NEW_USER.SIGNUP_FAIL);
  call(handleUserSignupFail)
}
```
But this won't work as expected, because the  SIGNUP_FAIL require all its previous actions to happen.
How do I handle this? Any suggestions, please.

**Comment by metagrover** [comment]
**tldr**: You can use `takeEvery`.

The video only discussed a normal scenario representing the idea that you can control the flow of action handling. Ideally, for auth, you can construct the flow something like this:

```js
function* handleSignup(username, password) {
  try {
    const user = yield call(Api.signup, username, password);
    yield put({ type: 'NEW_USER.SIGNUP_SUCCESS', user });
    // signup success can also trigger the welcome note for your use-case.
  }
  catch(error) {
    yield put({ type: 'NEW_USER.SIGNUP_FAIL', error });
  }
}
```

You can use `takeEvery` to watch for `NEW_USER.SIGNUP_FAIL` actions, or you can directly trigger `call(handleUserSignupFail)` inside the above catch block. Essentially ending up with desired flow of control.

Hope that helps.

---

Any alternate thoughts @divyanshu013?

**Comment by Puspendert** [comment]
Thanks. Earlier I was using the `takeLatest` and that was working fine. Just wanted to know your thoughts on this.

Also, may I please request you to have a look at an issue I am facing, with `axios` + `redux-saga`?
I followed the same code structure as you used in the tutorial, but instead of using `fetch` API, I used `axios`, but facing some issues.
[stackoverflow question](https://stackoverflow.com/questions/54302777/how-to-re-throw-the-catched-error-in-axios-catch-block)

**Comment by divyanshu013** [comment]
Yeah, ideally you should handle the failure inside your `catch` so that your watcher saga can always listen to the correct login request. Thanks for checking out our series. Cheers 🤝
